opentelemetry-sdk/src/trace/provider.rs

```diff
@@ -422,6 +422,25 @@ impl TracerProviderBuilder {
         TracerProviderBuilder { resource, ..self }
     }
 
+    /// Transforms the current [Resource] in this builder using a function.
+    ///
+    /// The transformed [Resource] represents the entity producing telemetry and
+    /// is associated with all [Tracer]s the [SdkTracerProvider] will create.
+    ///
+    /// By default, if this option is not used, the default [Resource] will be used.
+    ///
+    /// *Note*: Calls to this method are not additive, the returned resource will
+    /// completely replace the existing [Resource].
+    ///
+    /// [Tracer]: opentelemetry::trace::Tracer
+    pub fn transform_resource<F>(self, f: F) -> Self
```

Comment:
The type constraint isn't so long ... would it be cleaner inline? Something like

```suggestion
     pub fn transform_resource(self, f: impl FnOnce(Option<&Resource>) -> Resource) -> Self {
```